### PR TITLE
Allow method registration if handler implements Register

### PR DESCRIPTION
**Pull request by prashantv**
Currently, channel/subchannel only allow Register if the type
of the handler is the inbuilt channelHandler or handlerMap types.

This is quite inflexible, e.g., as we add new handler types in #831,
we still want to support registration without extending the type checks.

Instead of looking for a specific type, we can make the logic cleaner
by looking for the Register method. This also makes the channel's
Register method, and the subchannel's Register method more consistent.

This change allows users to specify their own custom handlers that can
support registration (via the same Register method).

### CODE REVIEW

**Review by AllenLuUber — APPROVED** [review]

**Comment by AllenLuUber** [comment]
I can confirm with change #831 is easier and a panicked test succeeded after rebasing on this.

File: channel.go [hunk 1]
```diff
@@ -425,12 +425,14 @@ type Registrar interface {
...
 //
 // Register panics if the channel was constructed with an alternate root
-// handler.
+// handler that does not support Register.
 func (ch *Channel) Register(h Handler, methodName string) {
-	if _, ok := ch.handler.(channelHandler); !ok {
-		panic("can't register handler when channel configured with alternate root handler")
+	r, ok := ch.handler.(registrar)
+	if !ok {
+		panic("can't register handler when channel configured with alternate root handler without Register method")
```
Comment (cinchurge) [inline channel.go:432]: seems like this will break existing callers - is this intended? do we want to make this optional rather than mandatory?

Comment (prashantv) [reply channel.go:432]: This should be more permissive than before, previously we panicked if the `handler` wasn't of the type `channelHandler`, but now we check for an interface (that `channelHandler` implements), so any case that worked previously should also work now. Any specific scenario you had in mind that now breaks?

Comment (cinchurge) [reply channel.go:432]: right - somehow i missed the added `Register()` method fo channelHandler. 😓  s'all good

**Review by cinchurge — APPROVED** [review]

**Pull request merged**